src/Compilers/VisualBasic/Portable/Symbols/UsedAssemblies.vb

```diff
@@ -0,0 +1,211 @@
+
+                            If retargetingAssembly IsNot Nothing Then
+                                usedAssemblies = retargetingAssembly.UnderlyingAssembly.DeclaringCompilation.GetCompleteSetOfUsedAssemblies(cancellationToken)
+                                If usedAssemblies IsNot Nothing Then
+                                    For Each underlyingDependency As AssemblySymbol In retargetingAssembly.UnderlyingAssembly.SourceModule.ReferencedAssemblySymbols
+                                        If Not underlyingDependency.IsLinked AndAlso usedAssemblies.Contains(underlyingDependency) Then
+                                            Dim dependency As AssemblySymbol = Nothing
+
+                                            If Not DirectCast(retargetingAssembly.Modules(0), RetargetingModuleSymbol).RetargetingDefinitions(underlyingDependency, dependency) Then
+                                                Debug.Assert(retargetingAssembly.Modules(0).ReferencedAssemblySymbols.Contains(underlyingDependency))
+                                                dependency = underlyingDependency
+                                            End If
+
+                                            AddUsedAssembly(dependency, stack)
+                                        End If
+                                    Next
+                                End If
+
+                                AddReferencedAssemblies(retargetingAssembly, includeMainModule:=False, stack)
+
+                                Continue While
+                            End If
+
+                            AddReferencedAssemblies(current, includeMainModule:=True, stack)
+                        End While
+
+                        stack.Free()
+                    End SyncLock
+                End If
+
+                If SourceAssembly.CorLibrary IsNot Nothing Then
+                    ' Add core library
+                    AddUsedAssembly(sourceAssembly.CorLibrary)
+                End If
+            End If
+
+            _usedAssemblyReferencesFrozen = True
+        End Sub
+
+        Friend Sub AddUsedAssemblies(assemblies As ICollection(Of AssemblySymbol))
+            If Not assemblies.IsNullOrEmpty() Then
+                For Each candidate In assemblies
+                    AddUsedAssembly(candidate)
+                Next
+            End If
+        End Sub
+
+        Friend Function AddUsedAssembly(assembly As AssemblySymbol) As Boolean
+
+            If assembly Is Nothing OrElse assembly Is SourceAssembly OrElse assembly.IsMissing Then
+                Return False
+            End If
+
+            If _lazyUsedAssemblyReferences Is Nothing Then
+                Interlocked.CompareExchange(_lazyUsedAssemblyReferences, New ConcurrentSet(Of AssemblySymbol)(), Nothing)
+            End If
+
+#If DEBUG Then
+            Dim wasFrozen As Boolean = _usedAssemblyReferencesFrozen
+#End If
+            Dim result As Boolean = _lazyUsedAssemblyReferences.Add(assembly)
```

Comment:
>result [](start = 16, length = 6)

Nit: a better name would be `add`, would make the assertion below easier to read at a glance.